### DISCUSSION TITLE
Merge or Diff changes of binary assets

### DISCUSSION THREAD

**Discussion by OptimisticRaccoon** [Sixze/ALS-Refactored#23] [Q&A] [answered]
Not really ALS-related, but how do you merge or diff changes of binary assets?

I added the Plugin as a git submodule, but it looks like Diff using the UE4 Git plugin doesn't work on submodules?
Pretty much every commit has some changes to binary assets to whom I have made significant changes, for some it's obvious, but having to manually double-check every Graph or Node in the AnimBP with a copy is not too sustainable.

**Comment by Sixze** [comment, marked as answer]
I don't use submodule for ALS.

I have 2 projects: the first is just a working copy of the ALS repository, which is only used to commit changes to GitHub, and the second is my private project (hosted on Azure DevOps) with a full copy of ALS (not a submodule) inside the `Plugins` folder.

So, I do all the ALS plugin work in my second project where I can easily diff the animation blueprint and other binary assets, and then, after the work is done, I just replace all the content in my first project with the content from the second project and then commit the changes to GitHub.

I propose the following rules when working with ALS in your project:
1. Do not use ALS as a submodule, but instead put a full copy of the ALS plugin in the `Plugins` folder.
2. Try to avoid making any modifications to the ALS content, and instead, if you want to change an asset, make a copy of it, place it somewhere in your project's `Content` folder, and edit that copy instead (for example in my project i made a copy of ABP_Als and added new overlay states to it).
3. When ALS has been updated, simply replace all ALS plugin files inside your project with files from the original repository.
4. Use diff in the Unreal editor if you want to know what exactly has changed inside binary assets, and then manually move changes to your asset copies if necessary.

It would be a real miracle if Epic finally added text serialization of assets so that they can be diffed and merged without the Unreal editor, but for now it is simply better to avoid merging and, if necessary, create copies of assets from the ALS plugin and change them if you do not want to deal with merge conflicts.

**Reply by emrehan98** [reply]
I'm not the one who asked the question but this is a pretty valuable answer. I will use that method, thanks!

**Reply by OptimisticRaccoon** [reply]
Definitely thanks for the in-depth answer, there was just one small detail I've missed. I thought the Diffing functionality was restricted to Source Control (Diff Against Depot) only. So I thought about rolling a Diff-anything, I strolled a bit through the source and noticed you already can do that by selecting 2 or more Assets x)

**Reply by Sixze** [reply]
Wow, I didn't know about that, really useful feature)

**Reply by AchimTuran** [reply]
Sadly the git plugin of the editor does not yet support submodules to create diffs. What we do is cloning the plugin with Git and add the folder to the ignore list. So you have two repositories one for you main project and one for the plugin. Sadly you can not use the editor diff feature. Would be very interestesting to know how much effort it would be to add this feature.

Maybe some of you might find this link interesting as well: https://www.stevestreeting.com/2020/08/09/my-ue4-vcs-setup---gitea---git---lfs---locking/ he did some changes to the git plugin to improve the workflow with Git. Sadly I did not yet had the time to check it out.

Additionally, we package the plugin and install it in the engine folder, which is a bit designer friendly. We also use the Workflow that we don't touch the plugin binary assets like BPs, AnimBPs... if changes are needed we copy them as well to the main project.
